overlord/devicestate/systems.go

```diff
@@ -176,3 +186,212 @@ func seededSystemFromModeenv() (*seededSystem, error) {
+		// we have in snap.Info, but getting it this way can be
+		// expensive as we need to compute the hash, try to find a
+		// better way
+		_, aRefs, err := seedwriter.DeriveSideInfo(sn.Path, f, db)
+		if err != nil {
+			if !asserts.IsNotFound(err) {
+				return nil, recoverySystemDir, err
+			} else if info.SnapID != "" {
+				// snap info from state must have come
+				// from the store, so it is unexpected
+				// if no assertions for it were found
+				return nil, recoverySystemDir, fmt.Errorf("internal error: no assertions for asserted snap with ID: %v", info.SnapID)
+			}
+		}
+		if err := w.SetInfo(sn, info); err != nil {
+			return nil, recoverySystemDir, err
+		}
+		sn.ARefs = aRefs
+	}
+
+	if err := w.InfoDerived(); err != nil {
+		return nil, recoverySystemDir, err
+	}
+
+	for {
+		// get the list of snaps we need in this iteration
+		toDownload, err := w.SnapsToDownload()
+		if err != nil {
+			return nil, recoverySystemDir, err
+		}
+		// which should be empty as all snaps should be accounted for
+		// already
+		if len(toDownload) > 0 {
+			which := make([]string, 0, len(toDownload))
+			for _, sn := range toDownload {
+				which = append(which, sn.SnapName())
+			}
+			return nil, recoverySystemDir, fmt.Errorf("internal error: need to download snaps: %v", strings.Join(which, ", "))
+		}
+
+		complete, err := w.Downloaded()
+		if err != nil {
+			return nil, recoverySystemDir, err
+		}
+		if complete {
+			logger.Noticef("snap processing complete")
+			break
+		}
+	}
+
+	for _, warn := range w.Warnings() {
+		logger.Noticef("WARNING: %s", warn)
+	}
+
+	unassertedSnaps, err := w.UnassertedSnaps()
+	if err != nil {
+		return nil, recoverySystemDir, err
+	}
+	if len(unassertedSnaps) > 0 {
+		locals := make([]string, len(unassertedSnaps))
+		for i, sn := range unassertedSnaps {
+			locals[i] = sn.SnapName()
+		}
+		logger.Noticef("system %q contains unasserted snaps %s", label, strutil.Quoted(locals))
+	}
+
+	copySnap := func(name, src, dst string) error {
+		if osutil.FileExists(dst) && strings.HasPrefix(dst, assertedSnapsDir+"/") {
+			// unasserted snaps are not shared
+			return nil
+		}
+		logger.Noticef("copying new seed snap %q from %v to %v", name, src, dst)
+		newFiles = append(newFiles, dst)
+		return osutil.CopyFile(src, dst, 0)
+	}
+	if err := w.SeedSnaps(copySnap); err != nil {
+		return newFiles, recoverySystemDir, err
+	}
+	if err := w.WriteMeta(); err != nil {
+		return newFiles, recoverySystemDir, err
+	}
+
+	bootSnaps, err := w.BootSnaps()
+	if err != nil {
+		return newFiles, recoverySystemDir, err
+	}
+	bootWith := &boot.RecoverySystemBootableSet{}
+	for _, sn := range bootSnaps {
+		switch sn.Info.Type() {
+		case snap.TypeKernel:
+			bootWith.Kernel = sn.Info
+			bootWith.KernelPath = sn.Path
+		case snap.TypeGadget:
+			bootWith.GadgetSnapOrDir = sn.Path
+		}
+	}
+	if err := boot.MakeRecoverySystemBootable(boot.InitramfsUbuntuSeedDir, recoverySystemDirInRootDir, bootWith); err != nil {
```

Comment:
see I imagine a world in which pretty much the entire code up to this point is some helper function inside seedwriter and that function just returns us the set of BootSnaps that we can use to build bootWith and pass that to MakeRecoverySystemBootable